packages/client/lib/sync/fetcher/accountfetcher.ts

```diff
@@ -0,0 +1,332 @@
+    origin: Buffer,
+    { accounts, proof }: { accounts: AccountData[]; proof: Buffer[] }
+  ): Promise<void> {
+    this.debug(
+      `verifyRangeProof accounts:${accounts.length} first=${short(accounts[0].hash)} last=${short(
+        accounts[accounts.length - 1].hash
+      )}`
+    )
+
+    for (let i = 0; i < accounts.length - 1; i++) {
+      // ensure the range is monotonically increasing
+      if (accounts[i].hash.compare(accounts[i + 1].hash) === 1) {
+        throw Error(
+          `Account hashes not monotonically increasing: ${i} ${accounts[i].hash} vs ${i + 1} ${
+            accounts[i + 1].hash
+          }`
+        )
+      }
+    }
+
+    const trie = new Trie({ db: new LevelDB() })
+    const keys = accounts.map((acc: any) => acc.hash)
+    const values = accounts.map((acc: any) => accountBodyToRLP(acc.body))
+    // convert the request to the right values
+    await trie.verifyRangeProof(stateRoot, origin, keys[keys.length - 1], keys, values, <any>proof)
+  }
+
+  private isMissingRightRange(
+    limit: Buffer,
+    { accounts, proof: _proof }: { accounts: AccountData[]; proof: Buffer[] }
+  ): boolean {
+    if (accounts.length > 0 && accounts[accounts.length - 1]?.hash.compare(limit) >= 0) {
+      return false
+    } else {
+      // TODO: Check if there is a proof of missing limit in state
+      return true
+    }
+  }
+
+  private getOrigin(job: Job<JobTask, AccountData[], AccountData>): Buffer {
+    const { task, partialResult } = job
+    const { first } = task
+    // Snap protocol will automatically pad it with 32 bytes left, so we don't need to worry
+    const origin = partialResult
+      ? bigIntToBuffer(bufferToBigInt(partialResult[partialResult.length - 1].hash) + BigInt(1))
+      : bigIntToBuffer(first)
+    return setLengthLeft(origin, 32)
+  }
+
+  private getLimit(job: Job<JobTask, AccountData[], AccountData>): Buffer {
+    const { task } = job
+    const { first, count } = task
+    const limit = bigIntToBuffer(first + BigInt(count) - BigInt(1))
+    return setLengthLeft(limit, 32)
+  }
+
+  /**
+   * Request results from peer for the given job.
+   * Resolves with the raw result
+   * If `undefined` is returned, re-queue the job.
+   * @param job
+   * @param peer
+   */
+  async request(
+    job: Job<JobTask, AccountData[], AccountData>
+  ): Promise<AccountDataResponse | undefined> {
+    const { peer } = job
+    const origin = this.getOrigin(job)
+    const limit = this.getLimit(job)
+
+    const rangeResult = await peer!.snap!.getAccountRange({
+      root: this.root,
+      origin,
+      limit,
+      bytes: BigInt(this.config.maxRangeBytes),
+    })
+
+    const peerInfo = `id=${peer?.id.slice(0, 8)} address=${peer?.address}`
+
+    // eslint-disable-next-line eqeqeq
+    if (rangeResult === undefined) {
+      return undefined
+    } else {
+      // validate the proof
+      try {
+        // verifyRangeProof will also verify validate there are no missed states between origin and
```

Comment:
@jochem-brouwer @scorbajio pls review if my understanding of proof verification is correct here w.r.t origin that using `origin` in verifyRangeProof (see the fn definition of this private fn) will validate that there are no `missed` accounts between origin and the first account recieved

Reply:
The way you are verifying makes sense to me. Could be useful to add some try-except blocks and log any errors in each function body.